Source/ROLib/Modules/ROLAnimateEngineHeat.cs

```diff
@@ -0,0 +1,216 @@
+                }
+            }
+            if (engineModule == null)
+            {
+                MonoBehaviour.print("ERROR: Could not locate engine by ID: " + engineID + " for part: " + part + " for ROLAnimateEngineHeat.  This will cause errors during gameplay.  Setting engine to first engine module (if present)");
+                if (engines.Length > 0) { engineModule = engines[0]; }
+            }
+        }
+
+        private void locateAnimatedTransforms()
+        {
+            List<Renderer> renderers = new List<Renderer>();
+            Transform[] animatedTransforms = part.transform.ROLFindChildren(meshName);
+            int len = animatedTransforms.Length;
+            for (int i = 0; i < len; i++)
+            {
+                renderers.AddRange(animatedTransforms[i].GetComponentsInChildren<Renderer>(false));
+            }
+            animatedRenderers = renderers.ToArray();
+            if (animatedRenderers == null || animatedRenderers.Length == 0) { print("ERROR: Could not locate any emissive meshes for name: " + meshName); }
+        }
+
+        private void updateHeat()
+        {
+            if (engineModule == null) { return; }
+
+            float emissivePercent = 0f;
+
+            if (!useThrottle)
+            {
+                //add heat from engine
+                if (engineModule.EngineIgnited && !engineModule.flameout && engineModule.currentThrottle > 0)
+                {
+                    float throttle = engineModule.currentThrottle;
+                    float heatIn = heatAccumulationCurve.Evaluate(throttle) * heatOutput * TimeWarp.fixedDeltaTime;
+                    currentHeat += heatIn;
+                }
+
+                //dissipate heat
+                float heatPercent = currentHeat / maxHeat;
+                if (currentHeat > 0f)
+                {
+                    float heatOut = heatDissipationCurve.Evaluate(heatPercent) * heatDissipation * TimeWarp.fixedDeltaTime;
+                    if (heatOut > currentHeat) { heatOut = currentHeat; }
+                    currentHeat -= heatOut;
+                }
+                if (currentHeat > maxStoredHeat) { currentHeat = maxStoredHeat; }
+
+                float mhd = maxHeat - draperPoint;
+                float chd = currentHeat - draperPoint;
+
+                if (chd < 0f) { chd = 0f; }
+                emissivePercent = chd / mhd;
+            }
+            else
+            {
+                emissivePercent = engineModule.currentThrottle;
+            }
+
+
+            if (emissivePercent > 1f) { emissivePercent = 1f; }
+            emissiveColor.r = redCurve.Evaluate(emissivePercent);
+            emissiveColor.g = greenCurve.Evaluate(emissivePercent);
+            emissiveColor.b = blueCurve.Evaluate(emissivePercent);
+            setEmissiveColors();
+        }
+
+        private void setEmissiveColors()
+        {
+            if (animatedRenderers != null)
+            {
+                bool rebuild = false;
```

Comment:
Possibly using LINQ in every FixedUpdate() would be less desirable, but this is also a very rolled-out form of `if (animateRenderers.Contains(null))` to find if you have any null renderers and just want to rebuild, using System.linq.

Reply:
I replaced the for with foreach, should be a bit more readable